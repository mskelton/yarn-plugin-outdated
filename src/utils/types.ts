export interface DependencyInfo {
  name: string
  current: string
  latest: string
  type: "dependencies" | "devDependencies"
}
