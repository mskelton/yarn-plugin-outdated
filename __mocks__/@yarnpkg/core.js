export const Cache = { find: jest.fn() }
export const Configuration = { find: jest.fn() }
export const Project = { find: jest.fn() }

export const structUtils = {
  stringifyIdent: ({ name, scope }) => (scope ? `@${scope}/${name}` : name),
}

export const formatUtils = {
  pretty: jest.fn(),
  prettyList: jest.fn(),
}

export const FormatType = {
  CODE: "CODE",
}
