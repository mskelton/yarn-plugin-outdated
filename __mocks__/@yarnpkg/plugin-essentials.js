export const suggestUtils = { fetchDescriptorFrom: jest.fn() }
