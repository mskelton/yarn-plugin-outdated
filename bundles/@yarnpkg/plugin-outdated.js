/* eslint-disable */
module.exports = {
name: "@yarnpkg/plugin-outdated",
factory: function (require) {
var plugin;plugin=(()=>{"use strict";var e={706:(e,t,r)=>{r.r(t),r.d(t,{default:()=>s});var n=r(747);const s={commands:[n.OutdatedCommand]}},742:(e,t,r)=>{Object.defineProperty(t,"__esModule",{value:!0}),t.DependencyFetcher=void 0;const n=r(850),s=r(687);t.DependencyFetcher=class{constructor(e,t,r){this.project=e,this.workspace=t,this.cache=r}async fetch(e,t){const r=await n.suggestUtils.fetchDescriptorFrom(e,t,{cache:this.cache,preserveModifier:e.range,project:this.project,workspace:this.workspace});return s.parseVersion(null===r?e.range:r.range)}}},732:(e,t,r)=>{Object.defineProperty(t,"__esModule",{value:!0}),t.DependencyTable=void 0;const n=r(966),s=r(513),o=["name","current","wanted","latest","workspace","type"];t.DependencyTable=class{constructor(e,t,r){this.configuration=e,this.dependencies=t,this.extraColumns=r,this.sizes=null,this.headers={current:"Current",latest:"Latest",name:"Package",type:"Package Type",wanted:"Wanted",workspace:"Workspace"}}print(){this.sizes=this.getColumnSizes(),this.printHeader(),this.dependencies.forEach(e=>{var t,r;const o=s.parse(e.current),a=s.parse(e.latest),i=a.major>o.major?"red":a.minor>o.minor?"yellow":"green";this.printRow({current:e.current.padEnd(this.sizes.current),latest:e.latest.padEnd(this.sizes.latest),name:n.formatUtils.pretty(this.configuration,e.name.padEnd(this.sizes.name),i),type:e.type.padEnd(this.sizes.type),wanted:null===(t=e.wanted)||void 0===t?void 0:t.padEnd(this.sizes.wanted),workspace:null===(r=e.workspace)||void 0===r?void 0:r.padEnd(this.sizes.workspace)})})}getColumnSizes(){const e={current:this.headers.current.length,latest:this.headers.latest.length,name:this.headers.name.length,type:this.headers.type.length,wanted:this.headers.wanted.length,workspace:this.headers.workspace.length};for(const t of this.dependencies)for(const[r,n]of Object.entries(t)){const t=e[r],s=(n||"").length;e[r]=t>s?t:s}return e}formatColumnHeader(e){return n.formatUtils.pretty(this.configuration,this.headers[e].padEnd(this.sizes[e]),"bold")}printHeader(){this.printRow({current:this.formatColumnHeader("current"),latest:this.formatColumnHeader("latest"),name:this.formatColumnHeader("name"),type:this.formatColumnHeader("type"),wanted:this.formatColumnHeader("wanted"),workspace:this.formatColumnHeader("workspace")})}printRow(e){const t=o.filter(e=>{var t;return null===(t=this.extraColumns[e])||void 0===t||t}).map(t=>e[t]).join("   ");console.log(t)}}},747:function(e,t,r){var n=this&&this.__decorate||function(e,t,r,n){var s,o=arguments.length,a=o<3?t:null===n?n=Object.getOwnPropertyDescriptor(t,r):n;if("object"==typeof Reflect&&"function"==typeof Reflect.decorate)a=Reflect.decorate(e,t,r,n);else for(var i=e.length-1;i>=0;i--)(s=e[i])&&(a=(o<3?s(a):o>3?s(t,r,a):s(t,r))||a);return o>3&&a&&Object.defineProperty(t,r,a),a};Object.defineProperty(t,"__esModule",{value:!0}),t.OutdatedCommand=void 0;const s=r(594),o=r(966),a=r(42),i=r(513),c=r(742),d=r(732),p=r(687);class l extends s.BaseCommand{constructor(){super(...arguments),this.all=!1,this.wanted=!1,this.json=!1}async execute(){const{cache:e,configuration:t,project:r,workspace:n}=await this.loadProject(),s=new c.DependencyFetcher(r,n,e),o=this.getWorkspaces(r,n),a=this.getDependencies(o),i=await this.getOutdatedDependencies(s,a);this.json?console.log(JSON.stringify(i)):new d.DependencyTable(t,i,{wanted:this.wanted,workspace:this.all}).print()}async loadProject(){const e=await o.Configuration.find(this.context.cwd,this.context.plugins),[t,{project:r,workspace:n}]=await Promise.all([o.Cache.find(e),o.Project.find(e,this.context.cwd)]);if(!n)throw new s.WorkspaceRequiredError(r.cwd,this.context.cwd);return{cache:t,configuration:e,project:r,workspace:n}}getWorkspaces(e,t){return this.all?e.workspaces:[t]}getDependencies(e){const t=[],r=["dependencies","devDependencies"];for(const n of e)for(const e of r)for(const r of n.manifest[e].values())t.push({dependencyType:e,descriptor:r,workspace:n});return t}async getOutdatedDependencies(e,t){const r=t.map(async({dependencyType:t,descriptor:r,workspace:n})=>{const s=i.valid(r.range)?"^"+r.range:r.range,[a,c]=await Promise.all([e.fetch(r,"latest"),this.wanted?e.fetch(r,s):void 0]);return{current:p.parseVersion(r.range),latest:a,name:o.structUtils.stringifyIdent(r),type:t,wanted:c,workspace:this.all?n.computeCandidateName():void 0}});return(await Promise.all(r)).filter(e=>i.neq(e.current,e.latest)).sort((e,t)=>e.name.localeCompare(t.name))}}n([a.Command.Boolean("-a,--all")],l.prototype,"all",void 0),n([a.Command.Boolean("-w,--wanted")],l.prototype,"wanted",void 0),n([a.Command.Boolean("--json")],l.prototype,"json",void 0),n([a.Command.Path("outdated")],l.prototype,"execute",null),t.OutdatedCommand=l},687:(e,t,r)=>{Object.defineProperty(t,"__esModule",{value:!0}),t.parseVersion=void 0;const n=r(513);t.parseVersion=e=>n.valid(n.coerce(e))},594:e=>{e.exports=require("@yarnpkg/cli")},966:e=>{e.exports=require("@yarnpkg/core")},850:e=>{e.exports=require("@yarnpkg/plugin-essentials")},42:e=>{e.exports=require("clipanion")},513:e=>{e.exports=require("semver")}},t={};function r(n){if(t[n])return t[n].exports;var s=t[n]={exports:{}};return e[n].call(s.exports,s,s.exports,r),s.exports}return r.n=e=>{var t=e&&e.__esModule?()=>e.default:()=>e;return r.d(t,{a:t}),t},r.d=(e,t)=>{for(var n in t)r.o(t,n)&&!r.o(e,n)&&Object.defineProperty(e,n,{enumerable:!0,get:t[n]})},r.o=(e,t)=>Object.prototype.hasOwnProperty.call(e,t),r.r=e=>{"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},r(706)})();
return plugin;
}
};