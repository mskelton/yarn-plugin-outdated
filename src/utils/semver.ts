import * as semver from "semver"

export const parseVersion = (version: string) =>
  semver.valid(semver.coerce(version))!
