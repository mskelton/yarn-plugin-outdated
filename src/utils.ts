import { Manifest } from "@yarnpkg/core"
import { fromUrl } from "hosted-git-info"
import semver from "semver"

export const truthy = Boolean as unknown as <T>(
  arg: T | undefined | null | false
) => arg is T

export function getHomepageURL({ raw: manifest }: Manifest): string | null {
  const repo = manifest.repository
  const repoURL = manifest.homepage
    ? manifest.homepage
    : typeof repo === "string"
    ? repo
    : typeof repo === "object" && typeof repo.url === "string"
    ? repo.url
    : null

  const info = repoURL ? fromUrl(repoURL) : undefined
  const commitish = info?.committish ? `#${info.committish}` : ""

  return info
    ? `https://${info.domain}/${info.user}/${info.project}${commitish}`
    : repoURL
}

/**
 * Because some packages have a pre-release version as their `latest` version,
 * we need to first check if the latest version is a pre-release. If it is,
 * we compare the current and latest directly, otherwise we coerce the current
 * version to remove any pre-release identifiers to determine if it is outdated.
 */
export function isVersionOutdated(current: string, latest: string) {
  return semver.parse(latest)!.prerelease.length
    ? semver.lt(current, latest)
    : semver.lt(semver.coerce(current)!, latest)
}
