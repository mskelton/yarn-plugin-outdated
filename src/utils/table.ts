import { Configuration, formatUtils } from "@yarnpkg/core"
import * as semver from "semver"
import { DependencyInfo } from "./types"

const columns = ["name", "current", "latest", "type"] as const

export type TableColumn = typeof columns[number]

export const headers: Record<TableColumn, string> = {
  current: "Current",
  latest: "Latest",
  name: "Package",
  type: "Package Type",
}

// Iterates through the dependencies to determine the necessary size
// of each column.
function getColumnSizes(
  dependencies: DependencyInfo[]
): Record<TableColumn, number> {
  const sizes = {
    current: headers.current.length,
    latest: headers.latest.length,
    name: headers.name.length,
    type: headers.type.length,
  }

  for (const dependency of dependencies) {
    for (const [key, value] of Object.entries(dependency)) {
      const acc = sizes[key as TableColumn]
      const cur = value.length

      sizes[key as TableColumn] = acc > cur ? acc : cur
    }
  }

  return sizes
}

function printHeader(
  configuration: Configuration,
  sizes: Record<TableColumn, number>
) {
  const format = (key: TableColumn) =>
    formatUtils.pretty(configuration, headers[key].padEnd(sizes[key]), "bold")

  printRow({
    current: format("current"),
    latest: format("latest"),
    name: format("name"),
    type: format("type"),
  })
}

function printRow(row: Record<TableColumn, string>) {
  console.log(columns.map((column) => row[column]).join("   "))
}

export function printTable(
  configuration: Configuration,
  dependencies: DependencyInfo[]
) {
  const sizes = getColumnSizes(dependencies)
  printHeader(configuration, sizes)

  dependencies.forEach((dep) => {
    const current = semver.parse(dep.current)!
    const latest = semver.parse(dep.latest)!

    const color =
      latest.major > current.major
        ? "red"
        : latest.minor > current.minor
        ? "yellow"
        : "green"

    printRow({
      current: dep.current.padEnd(sizes.current),
      latest: dep.latest.padEnd(sizes.latest),
      name: formatUtils.pretty(
        configuration,
        dep.name.padEnd(sizes.name),
        color
      ),
      type: dep.type.padEnd(sizes.type),
    })
  })
}
