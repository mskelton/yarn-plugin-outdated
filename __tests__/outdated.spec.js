import { formatUtils } from "../__mocks__/@yarnpkg/core"
import { run } from "./utils"

const candidates = {
  "@test-scope/package-a": "^3.4.0",
  "package-a": "1.1.0",
  "package-b": "^1.3.0",
  "package-c": "~1.2.1",
  "package-d": "^2.0.0",
}

describe("yarn outdated", () => {
  it("should show outdated dependencies", async () => {
    const output = await run("outdated", {
      candidates,
      manifest: {
        dependencies: {
          "package-a": "1.1.0",
          "package-b": "^1.1.1",
          "package-c": "~1.2.0",
        },
      },
    })

    expect(output).toMatchSnapshot()
  })

  it("should show outdated devDependencies", async () => {
    const output = await run("outdated", {
      candidates,
      manifest: {
        devDependencies: {
          "package-b": "^1.1.1",
          "package-c": "~1.2.1",
        },
      },
    })

    expect(output).toMatchSnapshot()
  })

  it("should highlight package names and version ranges", async () => {
    formatUtils.pretty.mockImplementation(
      (_, string, format) => `{${format}}${string}{${format}}`
    )

    const output = await run("outdated", {
      candidates,
      manifest: {
        dependencies: {
          "package-a": "1.1.0",
          "package-b": "^1.1.1",
          "package-c": "~1.2.0",
          "package-d": ">= 1",
        },
      },
    })

    expect(output).toMatchSnapshot()
  })

  it("should display empty state if no dependencies are outdated", async () => {
    const output = await run("outdated", {
      candidates,
      manifest: {
        dependencies: {
          "package-a": "1.1.0",
        },
      },
    })

    expect(output).toBe("âœ¨ All your dependencies are up to date!\n")
  })

  it("should ignore non-semver ranges", async () => {
    const output = await run("outdated", {
      candidates,
      manifest: {
        devDependencies: {
          "package-b": "^1.1.1",
          "package-c": "next",
          "package-d": "workspace:^1.2.3",
        },
      },
    })

    expect(output).toMatchSnapshot()
  })

  it("should properly format scoped dependencies", async () => {
    const output = await run("outdated", {
      candidates,
      manifest: {
        dependencies: {
          "@test-scope/package-a": "^3.0.1",
        },
        devDependencies: {
          "package-a": "^1.0.1",
        },
      },
    })

    expect(output).toMatchSnapshot()
  })
})
