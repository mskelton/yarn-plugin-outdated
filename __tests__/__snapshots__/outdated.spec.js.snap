// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yarn outdated should highlight package names and version ranges 1`] = `
"{bold}Package  {bold}   {bold}Current{bold}   {bold}Latest{bold}   {bold}Package Type{bold}
{yellow}package-b{yellow}   1.1.1     1.{bold}{yellow}3.0 {yellow}{bold}   dependencies
{green}package-c{green}   1.2.0     1.2.{bold}{green}1 {green}{bold}   dependencies
{red}package-d{red}   1.0.0     {bold}{red}2.0.0 {red}{bold}   dependencies
"
`;

exports[`yarn outdated should ignore non-semver ranges 1`] = `
"Package     Current   Latest   Package Type   
package-b   1.1.1     1.3.0    devDependencies
package-d   1.2.3     2.0.0    devDependencies
"
`;

exports[`yarn outdated should properly format scoped dependencies 1`] = `
"Package                 Current   Latest   Package Type   
@test-scope/package-a   3.0.1     3.4.0    dependencies   
package-a               1.0.1     1.1.0    devDependencies
"
`;

exports[`yarn outdated should show outdated dependencies 1`] = `
"Package     Current   Latest   Package Type
package-b   1.1.1     1.3.0    dependencies
package-c   1.2.0     1.2.1    dependencies
"
`;

exports[`yarn outdated should show outdated devDependencies 1`] = `
"Package     Current   Latest   Package Type   
package-b   1.1.1     1.3.0    devDependencies
"
`;
